[Scikit Learn Algorithm Map](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html)

### Choosing the Right Model for our Data on the basis of Scikit Learn Algorithm Map. 

### Scikit Learn uses Estimator as another term for Machine Learning Model or Algorithm.

### Regression : Predicting Continuous Number.
-------------------------------------------------------------------------------------------------------------------------------
**Import the Required Libraries**

In [1]:
import pandas as pd             # Data Manipulation and Analysis.
import numpy as np              # Multi Dimensional Arrays and Matrices and Mathematical Functions.
import matplotlib.pyplot as plt # Data Visualizaton.

**Import Data Set**

In [2]:
from sklearn.datasets import load_boston  # Data Set Imports as a Dictionary.

Data = load_boston()

In [3]:
print(Data.data) # The Actual Data 

[[6.3200e-03 1.8000e+01 2.3100e+00 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [2.7310e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [2.7290e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [6.0760e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0959e-01 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [4.7410e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 7.8800e+00]]


In [4]:
print(Data.feature_names) # Name of Features / Columns / Variables of the Data

['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


**Convert Data to Data Frame so that we can perform Data Analysis and Manipulation to Create Machine Learning Model.**

In [5]:
Boston = pd.DataFrame(data = Data.data, columns = Data.feature_names)

Boston['Target'] = Data.target # The Target Variable that is to be Predicted.

In [6]:
Boston.head()  # The Data Frame is Ready

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  Target  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

**Check the Information of Data Set : Data Type, Total Number of Data Elements, Total Features in our Data Set.**

In [7]:
print(Boston.info())  # Data Type of All the Features are Numeric. 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
CRIM       506 non-null float64
ZN         506 non-null float64
INDUS      506 non-null float64
CHAS       506 non-null float64
NOX        506 non-null float64
RM         506 non-null float64
AGE        506 non-null float64
DIS        506 non-null float64
RAD        506 non-null float64
TAX        506 non-null float64
PTRATIO    506 non-null float64
B          506 non-null float64
LSTAT      506 non-null float64
Target     506 non-null float64
dtypes: float64(14)
memory usage: 55.5 KB
None


### According to the Scikit Learn Map, We will Try Ridge Regression.

**Ridge Regression**

In [8]:
from sklearn.linear_model import Ridge

Ridge_Model = Ridge() # Instantiate Ridge Object for Training the Model.

**Split the Data Set into Training Set and Testing Set**

In [9]:
X = Boston.drop('Target', axis = 'columns') # Independent Features.

Y = Boston['Target']                        # Dependent Feature : Labels.

In [10]:
from sklearn.model_selection import train_test_split

np.random.seed(101)

X_Train, X_Test, Y_Train, Y_Test = train_test_split(X, Y, test_size = 0.3) 

# Test Size is used to Specify the Proportion we have to set for out Testing Set (0.3 is 30% of Data as Testing Set)

In [11]:
print('70% Training Set :',X_Train.shape,':','30% Testing Set :',X_Test.shape)

print('70% Training Set :',Y_Train.shape,'   :','30% Testing Set :',Y_Test.shape)

# X_Train : The Training Data Set                          

# Y_Train : The Set of Labels / Target Variable to all the Data in X_Train    

# X_Test : The Testing Data Set.

# Y_Test : The Set of Labels / Target variable for Comparison after Training the Model.  

70% Training Set : (354, 13) : 30% Testing Set : (152, 13)
70% Training Set : (354,)    : 30% Testing Set : (152,)


**Fit Model : Training the Model on our Data Set, Finding Patterns.**

In [12]:
Ridge_Model.fit(X_Train, Y_Train) # Train the Model using Training Data Sets.

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

**Check the Accuracy / Score of the Ridge Model on Test Data.**

In [13]:
print('Accuracy of Ridge Model :',Ridge_Model.score(X_Test, Y_Test)) # Ridge Model gave Accuracy of 67%

Accuracy of Ridge Model : 0.7102644986136442


### How do we Improve this Score ?

### What if Ridge isn't Working ?

**According to The Scikit Learn Algorithms Cheatsheet, Let's Try Random Forest Regressor.**

**Import Random Forest Regressor**

In [14]:
from sklearn.ensemble import RandomForestRegressor

RFR_Model = RandomForestRegressor(n_estimators=100)

In [15]:
RFR_Model.fit(X_Train, Y_Train)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=100,
                      n_jobs=None, oob_score=False, random_state=None,
                      verbose=0, warm_start=False)

**Check the Accuracy of Random Forest Regressor Model on Test Data.**

In [16]:
print('Accuracy of Random Forest Regressor Model :',RFR_Model.score(X_Test, Y_Test))

Accuracy of Random Forest Regressor Model : 0.867576086209281


#### Random Forest Regressor Model increased the Accuracy by 15% over the Ridge Model.

#### We can Improve the Accuracy by changing Train and Test Data Size and by changing the Number of Estimators.

#### Try at Your End it will Definitely Improve your Model.